### PR TITLE
Create Broadcast Client 2

### PR DESCRIPTION
**Pull request by CapCap**
## Motivation

(Write your motivation for proposed changes here.)

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?
yes

## Test Plan
Tests TBD

### CODE REVIEW

**Comment by libra-action** [comment]
**This PR may have modified JSON-RPC server code.**

Breaking changes policy:

1. Chaning JSON-RPC method signatures or removing/modifying fields
   in the response in /V1 will not be accepted.
2. Adding fields is ok.
3. Adding a new JSON-RPC method is ok.

Please ensure you have also done the following to update the docs:

1. Update json-rpc/API-CHANGELOG.md.
2. Add/update type/method documentation under json-rpc/docs.

**Comment by libra-action** [comment]
**LSR dependency change summary**:
```
target packages:
  M byteorder 1.4.3 (transitive third-party, crates.io)
    * version upgraded from 1.4.2
    * (unchanged features: i128)
  M libc 0.2.89 (transitive third-party, crates.io)
    * version upgraded from 0.2.88
    * (unchanged features: default, std)

host packages:
  M syn 1.0.64 (direct third-party, crates.io)
    * version upgraded from 1.0.62
    * (unchanged features: clone-impls, default, derive, extra-traits, full, parsing, printing, proc-macro, quote, visit)
```

**Comment by libra-action** [comment]
**release binaries dependency change summary**:
```
target packages:
  A move-stdlib 0.1.0 (workspace, path 'language/move-stdlib')
    * features: [none]
  M byteorder 1.4.3 (direct third-party, crates.io)
    * version upgraded from 1.4.2
    * (unchanged features: default, i128, std)
  M serde-generate 0.19.1 (direct third-party, crates.io)
    * version upgraded from 0.19.0
    * (unchanged features: [none])
  M libc 0.2.89 (transitive third-party, crates.io)
    * version upgraded from 0.2.88
    * (unchanged features: align, default, std)
  M regex-syntax 0.6.23 (transitive third-party, crates.io)
    * version upgraded from 0.6.22
    * (unchanged features: default, unicode, unicode-age, unicode-bool, unicode-case, unicode-gencat, unicode-perl, unicode-script, unicode-segment)

host packages:
  M syn 1.0.64 (direct third-party, crates.io)
    * version upgraded from 1.0.62
    * (unchanged features: clone-impls, default, derive, extra-traits, fold, full, parsing, printing, proc-macro, quote, visit, visit-mut)
  M libc 0.2.89 (transitive third-party, crates.io)
    * version upgraded from 0.2.88
    * (unchanged features: default, std)
  M regex-syntax 0.6.23 (transitive third-party, crates.io)
    * version upgraded from 0.6.22
    * (unchanged features: default, unicode, unicode-age, unicode-bool, unicode-case, unicode-gencat, unicode-perl, unicode-script, unicode-segment)
```

**Comment by libra-action** [comment]
**LEC dependency change summary**:
```
target packages:
  M byteorder 1.4.3 (transitive third-party, crates.io)
    * version upgraded from 1.4.2
    * (unchanged features: i128)
  M libc 0.2.89 (transitive third-party, crates.io)
    * version upgraded from 0.2.88
    * (unchanged features: align, default, std)

host packages:
  M syn 1.0.64 (direct third-party, crates.io)
    * version upgraded from 1.0.62
    * (unchanged features: clone-impls, default, derive, extra-traits, full, parsing, printing, proc-macro, quote, visit, visit-mut)
```

File: sdk/client/src/broadcast_client.rs [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        // All of them returned an error, so return the first error
+        if results.iter().all(|x| x.is_err()) {
+            if let Err(e) = results.swap_remove(0) {
+                return Err(e);
+            }
+        }
+        Ok(results.into_iter().filter_map(|x| x.ok()).collect())
+    }
+
+    pub async fn submit(&self, txn: &SignedTransaction) -> Result<Vec<Response<()>>> {
```
Comment (bmwill) [inline sdk/client/src/broadcast_client.rs:60]: At the end of the day we should only ever return 1 response, not all of the responses. So the logic in collect results should probably try to select just 1 to return.

File: sdk/client/src/broadcast_client.rs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+    pub async fn get_state_proof(&self, from_version: u64) -> Result<Vec<Response<StateProofView>>> {
+        let futures = self
+            .random_clients()
+            .map(|client| client.get_state_proof(from_version));
+        let results = join_all(futures).await;
+        collect_results(results)
+    }
+
+    pub async fn get_account_state_with_proof(&self,
```
Comment (bmwill) [inline sdk/client/src/broadcast_client.rs:167]: The formatting on these looks really weird, maybe you need to run rustfmt on it?

File: sdk/client/src/broadcast_client.rs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+            .random_clients()
+            .map(|client| client.get_events_with_proofs(key,
+                                                        start_seq,
+                                                        limit, ));
+        let results = join_all(futures).await;
+        collect_results(results)
+    }
+}
+
+fn collect_results<T>(mut results: Vec<Result<Response<T>>>) -> Result<Vec<Response<T>>> {
```
Comment (bmwill) [inline sdk/client/src/broadcast_client.rs:205]: Take a look at the old `select_highest_ledger_version_response` for how to select just 1 of the N responses to return.

File: sdk/client/Cargo.toml [hunk 4]
```diff
@@ -20,6 +20,8 @@ bcs = "0.1.2"
 hex = "0.4.3"
 serde = { version = "1.0.124", features = ["derive"] }
 serde_json = "1.0.64"
+futures = "0.3.12"
+rand = "0.8.3"
```
Comment (bmwill) [inline sdk/client/Cargo.toml:24]: This new broadcasting client and these deps need to only be turned on when the "async" feature is requested if we have this code live in this crate. Actually to make things easier on you (not needing to worry about this config stuff) mind moving this code into the old `json-rpc-client` crate?

File: sdk/client/src/broadcast_client.rs [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+    // All of them returned an error, so return the first error
+    if results.iter().all(|x| x.is_err()) {
+        if let Err(e) = results.swap_remove(0) {
+            return Err(e);
+        }
+    }
+    if let Some(result) = results.into_iter().filter_map(|x| x.ok()).nth(0) {
+        return Ok(result);
+    }
+    unreachable!()
```
Comment (bmwill) [inline sdk/client/src/broadcast_client.rs:290]: You should avoid using `unreachable` and prefer to restructure the code. Also take a look at https://github.com/diem/diem/blob/main/client/json-rpc/src/async_client/http_client.rs#L371 because we shouldn't just pick the first ok response, but rather the most up to date one.

File: sdk/client/src/broadcast_client.rs [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+
+    pub fn random_clients(&self) -> SliceChooseIter<'_, [Client], Client> {
+        self.clients
+            .choose_multiple(&mut rand::thread_rng(), self.num_parallel_requests)
+    }
+
+    pub async fn batch(
+        &self,
+        requests: Vec<MethodRequest>,
+    ) -> Result<Vec<Vec<Result<Response<MethodResponse>>>>> {
```
Comment (bmwill) [inline sdk/client/src/broadcast_client.rs:117]: This still seems overly nested here.

File: sdk/client/src/broadcast_client.rs [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+
+    pub async fn wait_for_transaction(
+        &self,
+        address: AccountAddress,
+        seq: u64,
+        expiration_time_secs: u64,
+        txn_hash: &str,
+        timeout: Option<Duration>,
+        delay: Option<Duration>,
+    ) -> Result<Response<TransactionView>, WaitForTransactionError> {
```
Comment (bmwill) [inline sdk/client/src/broadcast_client.rs:69]: So this is sort of messy...I'm going to make an executive decision and say that you should just drop the wait_for functions for now. If AOS needs them we can add them later.

File: client/json-rpc/src/async_client/http_client.rs [hunk 8]
```diff
@@ -387,12 +387,7 @@ impl BroadcastHttpClient {
         if rpc_resp.iter().all(|x| x.is_err()) {
             return rpc_resp.swap_remove(0);
         }
-        // Filter out all the errors and choose the JsonRpcResponse with the highest diem_ledger_version and return that
-        let mut rpc_resp: Vec<JsonRpcResponse> =
-            rpc_resp.into_iter().filter_map(|x| x.ok()).collect();
-        rpc_resp.sort_by(|x, y| y.diem_ledger_version.cmp(&x.diem_ledger_version));
-        Ok(rpc_resp.swap_remove(0))
-    }
+kvteeulvebcgklvjduhebgvevtcgvgvt
```
Comment (bmwill) [inline client/json-rpc/src/async_client/http_client.rs:390]: accidental Yubikey?

File: sdk/client/src/client.rs [hunk 9]
```diff
@@ -82,32 +82,47 @@ impl Client {
...
-                    _ => return Err(WaitForTransactionError::TransactionExecutionFailed(txn)),
-                }
-                return Ok(Response::new(txn, state));
-            }
-
-            if let Some(state) = self.last_known_state() {
-                if expiration_time_secs <= state.timestamp_usecs / 1_000_000 {
-                    return Err(WaitForTransactionError::TransactionExpired);
-                }
+            if let Some(res) = self.check_account_transaction_exists(address, seq, expiration_time_secs, txn_hash).await {
```
Comment (bmwill) [inline sdk/client/src/client.rs:85]: Since we're dropping the `wait_for` functions lets revert this change, and the one below.

Comment (bmwill) [reply sdk/client/src/client.rs:85]: Would you be able to revert the changes to this file?

File: sdk/client/src/lib.rs [hunk 10]
```diff
@@ -27,6 +27,7 @@ mod state;
 pub use state::State;
 
 mod retry;
+
```
Comment (bmwill) [inline sdk/client/src/lib.rs:30]: please remove this line

File: sdk/client/src/client.rs (hunk 9 above)
Comment (CapCap) [reply sdk/client/src/client.rs]: yes, sorry about the confusion

**Review by bmwill — APPROVED** [review]
/land
